Automatically generate API overview documentation

edersondisouza opened the pull request:
This PR should address #50922. It basically:

- ~~Generate \since tags from doxgen and git (it may not be perfect, but it's a fairly comprehensive start);~~
- Manually ~~fix~~ add some based on current information from doc/develop/api.overview.rst;
- Add \version (showing stability level) based on information from doc/develop/api.overview.rst;
- Use all new info (\since and \version) to generate the new API overview page, which is basically an improved version of doxygen own "Modules index" page.

More details of each step are on the commit messages. Two things that still stand out:
 - To parse the doxygen XML info, I've used its own "doxmlparser" Python lib, but it needs to be installed from the sources, as there's no package on PyPi. There's an open issue at doxygen repo about it: https://github.com/doxygen/doxygen/issues/8848
 - The new API overview page points to the doxygen documentation, not for any RST that may exist for it. Since the page now comes from doxygen, we could a) add a new `xrefitem` tag to reference it, but then we'd have doxygen pointing to RST and RST pointing to doxygen, which could be an issue; or b) have another file to map RST pages for some doxygen groups, but it would be another burden to maintain. Any thoughts about this?

Fixes #50922

edersondisouza committed:
doc: Generate develop/api/overview.rst API table from doxygen

A new extension, api_overview.py, is used to, leveraging doxygen's
Python module doxmlparser, parse the doxygen generated XML files.
All groups ('defgroup' and 'addtogroup' tags) are collected, alongside
their 'version' and 'since' info.

From there, a new Sphinx directive `api-overview-table` is populated,
including the name of the group, and if available, their 'since' and
'version' information.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

edersondisouza committed:
include/zephyr: Add 'version' and 'since' tag to groups

Based on information from doc/develop/api/overview.rst, add current
version for some groups representing APIs, following the following
table:

  - Experimental:   0.1.0
  - Unstable:       0.8.0
  - Stable:         1.0.0

Also based on doc/develop/api/overview.rst, add 'since' tag to the
groups.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

fgrandel reviewed (commented):
LGTM (IEEE 802.15.4 / gPTP headers - versions still missing but guess we can add that later)

gmarull commented on doc/CMakeLists.txt:
```diff
@@ -134,6 +144,26 @@ add_custom_target(
 
 set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${GEN_DEVICETREE_REST_SCRIPT})
 
+#-------------------------------------------------------------------------------
+# API overview
+
+set(GEN_API_OVERVIEW_REST_SCRIPT ${CMAKE_CURRENT_LIST_DIR}/_scripts/gen_api_overview_rest.py)
+
+add_custom_target(
+  api_overview
+  COMMAND ${CMAKE_COMMAND} -E env
+  ${PYTHON_EXECUTABLE} ${GEN_API_OVERVIEW_REST_SCRIPT}
+    --doxygen-xml-output-dir ${DOXY_OUT}/xml
+    ${DOCS_SRC_DIR}/develop/api
+  VERBATIM
+  USES_TERMINAL
+  COMMENT "Generating API overview documentation..."
+)
+
+set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${GEN_API_OVERVIEW_REST_SCRIPT})
+
+add_dependencies(api_overview doxygen)
```
-1 on reintroducing scriptery in docs build (it required significant effort to clean up the mess we had with scripts). Please look into how Sphinx extensions are used. You can also find useful examples in sdk-nrf repository.

henrikbrixandersen commented:
What problem are we solving here? What is the benefit of having a long list of when a _group_ of API functions were introduced?

nashif commented:
> What problem are we solving here? What is the benefit of having a long list of when a _group_ of API functions were introduced?

this replaces the old API table which was manually managed, it will also add stability level, based on version of each API derived automatically as well.

gmarull commented on include/zephyr/audio/dmic.h:
```diff
@@ -28,6 +29,8 @@
  *
  * @defgroup audio_dmic_interface Digital Microphone Interface
+ * @since 1.13.0
+ * @version 0.1.0
```
IMHO, API versions are redundant. Zephyr is used as a monolith, you can't choose which API version you want. You choose the OS version. Also, it is not clear how versioning works. I'm afraid we'll have a long stale list of "0.1.0", so just adding more confusion for the developers.

gmarull replied:
Another point, `@since` at an API level doesn't say anything useful to me because an API can change a lot since its initial addition. This should be used per-function, Python docs are a great example.

nashif replied:
> IMHO, API versions are redundant. Zephyr is used as a monolith, you can't choose which API version you want. You choose the OS version. Also, it is not clear how versioning works. I'm afraid we'll have a long stale list of "0.1.0", so just adding more confusion for the developers.

This is not about having someone select the version, it is about tracking API status and stability and breaking changes. We had this in textual form in a table and we agreeded to move this to the where the API is defined and track it using semantic versioning.

henrikbrixandersen replied:
To be clear, I am not opposed to adding the `@since` and `@version` tags (I believe this was already agreed upon previously). I am questioning the usability of having all of these in a long list in the documentation (as opposed to listing them for each individual API function in the API documentation).

nashif replied:
> To be clear, I am not opposed to adding the `@since` and `@version` tags (I believe this was already agreed upon previously). I am questioning the usability of having all of these in a long list in the documentation (as opposed to listing them for each individual API function in the API documentation).

well, they will be in the doxygen API docs and the headers and in sphinx and in other places, this page is generated automatically and we can decide how to display it and where to display any way we want. A long list that is up to date and current is 1000 times better than a short and a stale list.
Btw, another potential usage of this information is diffing the API versions between 2 releases and help identify API changes that can feed into the migration guide etc.

gmarull commented:
My opinion: I don't see how the new page helps developers. It's just a longer page that will take some more time (and RAM) to get rendered. Imagine, CAN API jumps from 1.0.0 to 2.0.0. What does this tell me, that it has breaking changes? Where are these explained? Also, who will control semantic versioning is properly handled in such long list of APIs/sub-APIs? And as said, `@since` at an API level is mostly redundant to me, unless `@since` is tied to the `@version` field.

IIRC there was a PR deleting that table, I think is what makes more sense. For a complete API table, one can already browse Doxygen pages.

edersondisouza commented:
v2:
  - Using Sphinx extensions to generate the API table;
  - Rebased.

kartben commented:
Looks like adding myself as a reviewer messed up the list of already lost reviewers, maybe because I was on the GH mobile app? Sorry about that - hopefully the bot will re-add folks next time a workflow runs

gmarull commented on doc/CMakeLists.txt:
```diff
@@ -39,6 +39,16 @@ if(NOT LATEX_PDFLATEX_FOUND OR NOT LATEXMK)
   message(WARNING "LaTeX components not found. PDF build will not be available.")
 endif()
 
+execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} -c "import doxmlparser"
+    RESULT_VARIABLE EXIT_CODE
+    OUTPUT_QUIET ERROR_QUIET
+)
+
+if (NOT ${EXIT_CODE} EQUAL 0)
+  message(FATAL_ERROR "'doxmlparser' Python package was not found")
+endif()
+
```
not needed, just update requirements

edersondisouza replied:
But `doxmlparser` is not on PyPi, so it wouldn't be installed and we'd just have the import failure - isn't that worse? Am I missing something?

mbolivar-ampere replied:
Then IMO this should not be merged as-is, because it's breaking the use case of "build the docs locally after installing requirements from pip"

carlescufi replied:
@edersondisouza if `doxmlparser` is not in PyPi then why did you add it to `doc/requirements.txt`?

edersondisouza replied:
Oh, it is now. I've sent a PR to doxygen to publish it to PyPi (https://github.com/doxygen/doxygen/pull/10328) that eventually got merged and it is on PyPi now (https://pypi.org/project/doxmlparser/).

gmarull commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,127 @@
+    api_overview_doxygen_xml_dir: Doxygen xml output directory
+    '''
+
+    def run(self):
+        raw_node = nodes.raw(text=self.env.api_overview_content,format='html')
+        return [raw_node]
+
+
+tail_piece = "        </table>"
+
+row = """\
+            <tr class="{css_class}">
+                <td>
+                    <span style="width:{indent}px;display:inline-block">&nbsp;</span>
+                    <a href="{url}">{title}</a>
+                </td>
+                <td>{version}</td>
+                <td>{since}</td>
+            </tr>"""
+
+preamble = """
+        <table class="docutils align-default">
+            <thead>
+                <tr class="row-odd">
+                    <th class="head">API</th>
+                    <th class="head">Version</th>
+                    <th class="head">Since</th>
+                </tr>
+            </thead>"""
```
please use docutils to build a table

gmarull commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,127 @@
+        elif sect.get_kind() == 'version':
+            version = sect.get_para()[0].get_valueOf_()
+
+    url = f"../../doxygen/html/{cdef.get_id()}.html"
```
this needs to be configurable

gmarull commented on doc/conf.py:
```diff
@@ -319,6 +320,9 @@
 
+# -- Options for zephyr.api_overview --------------------------------------
+
+api_overview_doxygen_xml_dir = "html/doxygen/xml"
```
this needs to be a function of `doxyrunner_outdir`

mbolivar-ampere commented on include/zephyr/devicetree.h:
```diff
@@ -27,6 +27,7 @@
 /**
  * @brief devicetree.h API
  * @defgroup devicetree Devicetree
+ * @since 2.0.0
```
The entire DT API was rewritten completely for v2.3.0; this is definitely one of the cases where the automation is not working (I see that's explained in the commit message), but it does mean this content is incorrect.

edersondisouza replied:
If the DT API was there on version 2.0.0, but we need to track it as v2.3.0 because of a rewrite, then all bets are off for automation =D
The idea of the automated search was to provide some base - which would have errors, that hopefully reviewers with more historical knowledge would help fix - if not on this PR, over some time after it gets in. It seems fairly good for more recent stuff, but yeah, older APIs have more history to them.
My feeling then is that the best way here is to drop the results from the automated search and only add the ones that were present on the original overview.rst - thoughts?

mbolivar-ampere replied:
> My feeling then is that the best way here is to drop the results from the automated search

Yeah, removing the automation makes sense to me

mbolivar-ampere commented on include/zephyr/devicetree.h:
```diff
@@ -2455,6 +2463,7 @@
 
 /**
  * @defgroup devicetree-generic-foreach "For-each" macros
+ * @since 2.3.0
```
There are problems in the other direction as well. These APIs did not show up in v2.3.0. DT_FOREACH_NODE() was not part of the original API rewrite.

mbolivar-ampere reviewed (dismissed):
A quick look at the DT side makes it clear to me that something is wrong with the automation.

mbolivar-ampere commented:
> A quick look at the DT side makes it clear to me that something is wrong with the automation.

Emphasis on *quick look*: my comments are not meant to be exhaustive. I am not saying "please fix just these issues and all will be well". I am saying "something looks quite wrong here based on a couple of examples I found while spot checking and IMO it should be investigated"

edersondisouza commented:
v3:
  - No more mandatory requiring `doxmlparser` module - if it's not there, a warning will be produced during documentation build, and the overview table will not be generated, but everything else will be built;
  - Using doxygen XML build dir from doxyrunner configuration;
  - Base URL fro doxygen is now a configurable option;
  - Overview table built from docutils classes instead of hardcoded HTML;
  - Use `black` to reformat the api_overview.py file.

fgrandel reviewed (commented):
LGTM

gmarull commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,179 @@
+
+logger = logging.getLogger(__name__)
+
+doxmlparser = None
+if importlib.util.find_spec("doxmlparser"):
+    doxmlparser = importlib.import_module("doxmlparser")
+
+    from doxmlparser.compound import DoxCompoundKind
+else:
+    logger.warning(
+        "Module `doxmlparser` not found. API overview table will not be generated"
+    )
```
If this package is not pip-installable, then you should parse Doxygen files yourself (it's not that complicated). Depending on something like this is a bad idea (no versioning, how to install, etc).

nashif replied:
we should just include this in our tree in scripts or docs/ or whatever.

gmarull replied:
doesn't sound like a good idea by looking at https://github.com/doxygen/doxygen/tree/master/addon/doxmlparser. Maybe they should just publish this package. I remember this parser being very slow, to the point I discarded using it for https://github.com/zephyrproject-rtos/zephyr/pull/39702 has this been checked?

nashif replied:
for whatever we do here, that should be fast enough I think. if this becomes a package, we can move to it, but "slowness" will not go away when in pypi

kartben replied:
```python
    import time
    s = time.perf_counter()
    groups = parse_xml_dir(doxygen_xml_dir)
    logger.info(f"XML parsing took {time.perf_counter() - s:.2f} seconds")
```

> XML parsing took 10.67 seconds

edersondisouza commented:
v4:
 - Include `doxmlparser` code directly.

nashif commented:
https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html

gmarull commented:
> v4:
> 
> * Include `doxmlparser` code directly.

can you work with upstream Doxygen to provide them with a CI workflow that publishes the package on PyPI instead? It should be fairly easy. An example workflow https://github.com/nrfconnect/doc-sphinx-ncs-theme/blob/main/.github/workflows/default.yml

edersondisouza commented:
v5:
  - Dropped patch that added `since` automatically from git log;
  - Fixed CI issues with doxmlparser code;
  - Rebased.

gmarull reviewed (dismissed):
-1 until doxmlparser is published to PyPI, seems like we just need to wait one doxygen release (should happen before Zephyr's next release according to their past releases)

nashif commented:
we can also look into using `pip install -i https://test.pypi.org/simple/ doxmlparser` until there is an official release to help with getting populating the versioning information and having something useful by the next release. We do not need updated doxygen version itself, just this python package anyways.

fgrandel commented on include/zephyr/net/ieee802154.h:
```diff
@@ -26,6 +26,8 @@ extern "C" {
 /**
  * @defgroup ieee802154 IEEE 802.15.4 and Thread APIs
+ * @since 1.0
+ * @version 0.8.0
```
According to https://docs.zephyrproject.org/latest/develop/api/overview.html, the current version of all IEEE 802.15.4 APIs is now (since release 3.5) `1.0.0` - also concerns the other `ieee802154*` files. Last time I reviewed this was not the case but now I think this should be adjusted.

nashif replied:
the API based on the latest docs in https://docs.zephyrproject.org/latest/develop/api/overview.html is 'unstable'. 0.8.0 means 'unstable', it should not be 1.0.0 until the API is stable.

@edersondisouza  please put some documentation at the top of the page explaining the versions and what they mean.

kartben replied:
> According to https://docs.zephyrproject.org/latest/develop/api/overview.html, the current version of all IEEE 802.15.4 APIs is now (since release 3.5) `1.0.0`

FWIW the last column in the table is meant to indicate the API was **introduced** in Zephyr 1.0.0, so that recent update to the table you're mentioning was probably bogus in the first place.

fgrandel reviewed (changes requested):
Sorry for being so late, but I wasn't available for a few days and the change that I refer to was only introduced in release 3.5

edersondisouza committed:
doc: Include doxmlparser on requirements.txt

It will be necessary to parse the doxygen XML files to extract `since`
and `version` from them.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza commented:
v6:
  - Instead of including doxmlparser code, use it from TestPyPI for now (will update to use PyPI as soon as it's available there);
  - Added a simple table to explain the meaning of column "version";
  - Rebased.

gmarull commented on doc/CMakeLists.txt:
```diff
@@ -140,6 +140,7 @@ set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${GEN_DEVICETREE_
 add_doc_target(
   html
   COMMAND ${CMAKE_COMMAND} -E env ${SPHINX_ENV}
+  PYTHONPATH=${ZEPHYR_BASE}/doc/_extensions/zephyr${SEP}$ENV{PYTHONPATH}
```
why is this needed?

edersondisouza replied:
Oops, leftover from when doxmlparser code was included.

gmarull commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,161 @@
+        return [self.env.api_overview_table]
+
+
+spacer = '<span style="width:{indent}px;display:inline-block">&nbsp;</span>'
```
looks hacky, and HTML only (important as some still insist on PDF)

kartben replied:
+1, please stick to standard nodes and/or rely on CSS for tweaking the look&feel.

edersondisouza replied:
Ok, will try to convey the hierarchy with `-` instead.

edersondisouza commented:
v7:
  - Removed astray `PYTHONPATH` from `CMakeLists.txt`;
  - Hierarchy of doxygen groups represented via `-` instead of space indentation, so it works on PDF;
  - Rebased.

edersondisouza commented:
CI failure is unrelated, quick fix at https://github.com/zephyrproject-rtos/zephyr/pull/64807

edersondisouza commented:
v9:
 - Rebased to get the fix for CI issue.

henrikbrixandersen reviewed (dismissed):
As mentioned before, I think the generated list of API versions is a bit too long to be useful, but the functionality found in this PR does seem to match what the TSC decided upon.

Thank you for putting in the work on establishing this.

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+
+
+class ApiOverview(SphinxDirective):
+    r"""
```
nit

```suggestion
    """
```

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+
+def parse_xml_dir(dir_name):
+    groups = []
+    root = doxmlparser.index.parse(f"{dir_name}/index.xml", True)
```
use joinpath

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+    root = doxmlparser.index.parse(f"{dir_name}/index.xml", True)
+    for compound in root.get_compound():
+        if compound.get_kind() == DoxCompoundKind.GROUP:
+            file_name = f"{dir_name}/{compound.get_refid()}.xml"
```
ditto

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+
+    thead = nodes.thead()
+    thead_row = nodes.row()
+    for header_name in ["API", "Version", "Since"]:
```
as per comment re version vs. status, this should become:

```suggestion
    for header_name in ["API", "Status", "Since"]:
```

kartben commented on doc/develop/api/overview.rst:
```diff
@@ -7,384 +7,18 @@ The table lists Zephyr's APIs and information about them, including their
-   * - API
-     - Status
-     - Version Introduced
+   * - Version
+     - Meaning
+   * - 0.1.0
+     - Experimental API
+   * - 0.8.0
+     - Unstable API
+   * - 1.0.0
+     - Stable API
```
BTW if I am missing something with regards to a previous TSC or Process WG decision that we explicitly want to set actual versions (and not just express api status) please let me know.

kartben replied:
And if we do think `@version` is the right thing to do, then we should revise this table (and the default values initially set as part of this PR) to be less set in stone (i.e. not all unstable APIs will forever be 0.8.0 and only one day move to 1.0.0 -- and again stay there forever, right?), and better aligned with semver. Something like:

```suggestion
   * - Version
     - Meaning
   * - 0.0.x
     - Experimental API
   * - 0.x.0
     - Unstable API
   * - x.0.0
     - Stable API
```

edersondisouza replied:
While I don't disagree, I'm not sure we could just skip over TSC decision here - regarding using "Version" as SemVer. But I like the `X` idea, maybe we could have `0.1.x`, `0.8.x` and `1.x.x`?

nashif replied:
this is going to be confusing, it apears as if those versions are static. We need more verbiage here and we need to stick to the semver spec, i.e. `<version core> ::= <major> "." <minor> "." <patch>`, so we cant use 0.0.x or an experimental API should start with `0.0.0` and only increase the patch level until it is unstable, which is not great. Now that I think about it, the original makes more sense. We just need to more language to clarify this, i.e. APIs might move the version up as they evolve and  `1.1.7` or whatever is also a stable API version. If the API moved from `1.1.7` to `2.0.0`, this means there was a breaking change, etc.

edersondisouza replied:
So, something like (maybe copy as is) the `@version` section from https://github.com/zephyrproject-rtos/zephyr/issues/50922?

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+    entry += inline
+    row_node += entry
+
+    # Finally, add version and since
```
when version (... or rather status? :)) is not set, should we by default show the same value as the parent group?

edersondisouza replied:
Not sure, as there's a risk of implying something is more stable than it is.

kartben reviewed (dismissed):
The looks of the table and the way the extension is implemented is a +1 from me. 
I am unsure about the current meaning put behind `@version` and how the current documentaiton seems to imply that 0.1.0, 0.8.0, and 1.0.0 are the only versions that might ever exist, and how they effectively "just" mean experimental/unstable/stable.

kartben commented:
BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/94a5a588-b258-47cb-83ee-6a2c6cc76d08)

edersondisouza commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -0,0 +1,158 @@
+
+
+def get_group(innergroup, all_groups):
+    return [
+        g
+        for g in all_groups
+        if g.get_compounddef()[0].get_id() == innergroup.get_refid()
+    ][0]
```
No match here would mean a `doxmlparser` bug (saying some group reference another, but there's no "another"). I can try with some `try`, but feels unnecessary.

edersondisouza commented:
> BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.
> 
> ![image](https://user-images.githubusercontent.com/128251/281756129-94a5a588-b258-47cb-83ee-6a2c6cc76d08.png)

Maybe - but that would imply some way of marking them internal - a custom tag, defining some directories as internal, etc.  Not sure if it belongs on this PR, or a future one.

kartben commented:
> > BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.
> > 
> > ![image](https://user-images.githubusercontent.com/128251/281756129-94a5a588-b258-47cb-83ee-6a2c6cc76d08.png)
> 
> Maybe - but that would imply some way of marking them internal - a custom tag, defining some directories as internal, etc.  Not sure if it belongs on this PR, or a future one.

Right but the reason I'm pointing it here is that there it is *not* showing in Doxygen documentation (since it starts with the Z_ prefix, that's the intended behavior as it's currently one of the ways to exclude internal stuff from the doc - https://github.com/zephyrproject-rtos/zephyr/blob/56f73bde0fb086fae38cfb5afb2f2500a340b39b/doc/zephyr.doxyfile.in#L1018), so I just want to make sure the mismatch is looked into.

edersondisouza commented:
> > > BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.
> > > ![image](https://user-images.githubusercontent.com/128251/281756129-94a5a588-b258-47cb-83ee-6a2c6cc76d08.png)
> > 
> > 
> > Maybe - but that would imply some way of marking them internal - a custom tag, defining some directories as internal, etc.  Not sure if it belongs on this PR, or a future one.
> 
> Right but the reason I'm pointing it here is that there it is _not_ showing in Doxygen documentation (since it starts with the Z_ prefix, that's the intended behavior as it's currently one of the ways to exclude internal stuff from the doc -
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/56f73bde0fb086fae38cfb5afb2f2500a340b39b/doc/zephyr.doxyfile.in#L1018
> 
> ), so I just want to make sure the mismatch is looked into.

Ha, that should make it easier, then =D
But that exclude list also excludes things that start with `_`, and that would exclude USB-C stuff. Guess I'll skip only when it starts with `Z_`.

kartben commented:
> > > > BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.
> > > > ![image](https://user-images.githubusercontent.com/128251/281756129-94a5a588-b258-47cb-83ee-6a2c6cc76d08.png)
> > > 
> > > 
> > > Maybe - but that would imply some way of marking them internal - a custom tag, defining some directories as internal, etc.  Not sure if it belongs on this PR, or a future one.
> > 
> > Right but the reason I'm pointing it here is that there it is _not_ showing in Doxygen documentation (since it starts with the Z_ prefix, that's the intended behavior as it's currently one of the ways to exclude internal stuff from the doc -
> > 
> > https://github.com/zephyrproject-rtos/zephyr/blob/56f73bde0fb086fae38cfb5afb2f2500a340b39b/doc/zephyr.doxyfile.in#L1018
> > 
> > ), so I just want to make sure the mismatch is looked into.
> 
> Ha, that should make it easier, then =D
> But that exclude list also excludes things that start with `_`, and that would exclude USB-C stuff. Guess I'll skip only when it starts with `Z_`.

Wait I think the reason why the cbprintf group is not showing might not be the Z_ (since the USB-C groups are showing just fine despite their _) after all but rather the fact that it's actually empty. If not too complicated, this might be worth replicating this behavior.

edersondisouza commented:
> > > > > BTW I spotted this "cbvprintf processing flags." group showing up, but it should be internal? Guess there might be others.
> > > > > ![image](https://user-images.githubusercontent.com/128251/281756129-94a5a588-b258-47cb-83ee-6a2c6cc76d08.png)
> > > > 
> > > > 
> > > > Maybe - but that would imply some way of marking them internal - a custom tag, defining some directories as internal, etc.  Not sure if it belongs on this PR, or a future one.
> > > 
> > > 
> > > Right but the reason I'm pointing it here is that there it is _not_ showing in Doxygen documentation (since it starts with the Z_ prefix, that's the intended behavior as it's currently one of the ways to exclude internal stuff from the doc -
> > > https://github.com/zephyrproject-rtos/zephyr/blob/56f73bde0fb086fae38cfb5afb2f2500a340b39b/doc/zephyr.doxyfile.in#L1018
> > > 
> > > ), so I just want to make sure the mismatch is looked into.
> > 
> > 
> > Ha, that should make it easier, then =D
> > But that exclude list also excludes things that start with `_`, and that would exclude USB-C stuff. Guess I'll skip only when it starts with `Z_`.
> 
> Wait I think the reason why the cbprintf group is not showing might not be the Z_ (since the USB-C groups are showing just fine despite their _) after all but rather the fact that it's actually empty. If not too complicated, this might be worth replicating this behavior (after having confirmed the theory)?

It seems the group is not empty: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/cbprintf.h#L267

But doxygen generates the group nevertheless, one just need to know where to look: https://docs.zephyrproject.org/latest/doxygen/html/group__Z__CBVPRINTF__PROCESS__FLAGS.html
However, it will be empty because I guess the internal symbols were excluded due the prefix.

This seems to be the only group starting with `Z_`, and skipping it shall be easy.

edersondisouza commented:
v9:
  - Don't use string formats to join paths or urls
  - Skip groups starting in `Z_`, as those shall be internal;
  - Improve error handling for missing groups;
  - Formatting issues;
  - Reword API version meaning description;
  - Rebased.

edersondisouza commented:
> ```
> -- MCUmgr 		
> ---- MCUmgr callback API 		
> ------ MCUmgr fs_mgmt callback API 		
> ------ MCUmgr img_mgmt callback API 		
> ------ MCUmgr os_mgmt callback API 		
> ------ MCUmgr settings_mgmt callback API 		
> ---- MCUmgr handler API 		
> ---- MCUmgr img_mgmt API 		
> ---- MCUmgr mgmt API 	1.0.0 	1.11
> ---- MCUmgr transport SMP API
> ```
> 
> This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.

Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?

nordicjm commented:
> > ```
> > -- MCUmgr 		
> > ---- MCUmgr callback API 		
> > ------ MCUmgr fs_mgmt callback API 		
> > ------ MCUmgr img_mgmt callback API 		
> > ------ MCUmgr os_mgmt callback API 		
> > ------ MCUmgr settings_mgmt callback API 		
> > ---- MCUmgr handler API 		
> > ---- MCUmgr img_mgmt API 		
> > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > ---- MCUmgr transport SMP API
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> 
> Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?

It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.

edersondisouza commented:
v10:
  - Fixed bug on joining urls;
  - Rebased.

edersondisouza commented:
> > > ```
> > > -- MCUmgr 		
> > > ---- MCUmgr callback API 		
> > > ------ MCUmgr fs_mgmt callback API 		
> > > ------ MCUmgr img_mgmt callback API 		
> > > ------ MCUmgr os_mgmt callback API 		
> > > ------ MCUmgr settings_mgmt callback API 		
> > > ---- MCUmgr handler API 		
> > > ---- MCUmgr img_mgmt API 		
> > > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > > ---- MCUmgr transport SMP API
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> > 
> > 
> > Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?
> 
> It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.

About the MCUmgr client, I don't see it here now: https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html
Maybe is missing a `defgroup` or `addtogroup` on docs?

In any case, this just shows the hierarchy of groups on doxygen, why would the way the APIs are grouped for documentation not work to show their versions? Moving forward, as more `@version` and `@since` tags are added to doxygen docs, the more complete the table will be, and hopefully, the less "awful" it'll look.

nordicjm commented:
> > > > ```
> > > > -- MCUmgr 		
> > > > ---- MCUmgr callback API 		
> > > > ------ MCUmgr fs_mgmt callback API 		
> > > > ------ MCUmgr img_mgmt callback API 		
> > > > ------ MCUmgr os_mgmt callback API 		
> > > > ------ MCUmgr settings_mgmt callback API 		
> > > > ---- MCUmgr handler API 		
> > > > ---- MCUmgr img_mgmt API 		
> > > > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > > > ---- MCUmgr transport SMP API
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> > > 
> > > 
> > > Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?
> > 
> > 
> > It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.
> 
> About the MCUmgr client, I don't see it here now: https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html Maybe is missing a `defgroup` or `addtogroup` on docs?

@jheiskan81 please address this

> In any case, this just shows the hierarchy of groups on doxygen, why would the way the APIs are grouped for documentation not work to show their versions? Moving forward, as more `@version` and `@since` tags are added to doxygen docs, the more complete the table will be, and hopefully, the less "awful" it'll look.

Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)

jheiskan81 commented:
> > > > > ```
> > > > > -- MCUmgr 		
> > > > > ---- MCUmgr callback API 		
> > > > > ------ MCUmgr fs_mgmt callback API 		
> > > > > ------ MCUmgr img_mgmt callback API 		
> > > > > ------ MCUmgr os_mgmt callback API 		
> > > > > ------ MCUmgr settings_mgmt callback API 		
> > > > > ---- MCUmgr handler API 		
> > > > > ---- MCUmgr img_mgmt API 		
> > > > > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > > > > ---- MCUmgr transport SMP API
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> > > > 
> > > > 
> > > > Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?
> > > 
> > > 
> > > It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.
> > 
> > 
> > About the MCUmgr client, I don't see it here now: https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html Maybe is missing a `defgroup` or `addtogroup` on docs?
> 
> @jheiskan81 please address this
> 
> > In any case, this just shows the hierarchy of groups on doxygen, why would the way the APIs are grouped for documentation not work to show their versions? Moving forward, as more `@version` and `@since` tags are added to doxygen docs, the more complete the table will be, and hopefully, the less "awful" it'll look.
> 
> Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)

@nordicjm & @edersondisouza https://github.com/zephyrproject-rtos/zephyr/pull/65289 Please check is this fixing

nordicjm commented:
> > > > > > ```
> > > > > > -- MCUmgr 		
> > > > > > ---- MCUmgr callback API 		
> > > > > > ------ MCUmgr fs_mgmt callback API 		
> > > > > > ------ MCUmgr img_mgmt callback API 		
> > > > > > ------ MCUmgr os_mgmt callback API 		
> > > > > > ------ MCUmgr settings_mgmt callback API 		
> > > > > > ---- MCUmgr handler API 		
> > > > > > ---- MCUmgr img_mgmt API 		
> > > > > > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > > > > > ---- MCUmgr transport SMP API
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> > > > > 
> > > > > 
> > > > > Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?
> > > > 
> > > > 
> > > > It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.
> > > 
> > > 
> > > About the MCUmgr client, I don't see it here now: https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html Maybe is missing a `defgroup` or `addtogroup` on docs?
> > 
> > 
> > @jheiskan81 please address this
> > > In any case, this just shows the hierarchy of groups on doxygen, why would the way the APIs are grouped for documentation not work to show their versions? Moving forward, as more `@version` and `@since` tags are added to doxygen docs, the more complete the table will be, and hopefully, the less "awful" it'll look.
> > 
> > 
> > Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)
> 
> @nordicjm & @edersondisouza #65289 Please check is this fixing a problem?

Yes, resolves that problem. @edersondisouza now with that, there will be an additional 3 entries, MCUmgr will have 13 entries when it needs 2!

edersondisouza commented:
> > > > > > > ```
> > > > > > > -- MCUmgr 		
> > > > > > > ---- MCUmgr callback API 		
> > > > > > > ------ MCUmgr fs_mgmt callback API 		
> > > > > > > ------ MCUmgr img_mgmt callback API 		
> > > > > > > ------ MCUmgr os_mgmt callback API 		
> > > > > > > ------ MCUmgr settings_mgmt callback API 		
> > > > > > > ---- MCUmgr handler API 		
> > > > > > > ---- MCUmgr img_mgmt API 		
> > > > > > > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > > > > > > ---- MCUmgr transport SMP API
> > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > >     
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >     
> > > > > > >   
> > > > > > > This looks awful, the whole thing needs to go and just have 2 MCUmgr versions, not 10 - MCUmgr and MCUmgr client (which, apparently, is not even showing right now). Auxdisplay is missing from the list.
> > > > > > 
> > > > > > 
> > > > > > Those come from `@defgroup` , `@addgroup` and `@ingroup` definitions - if they don't make sense, maybe refining the groups is the way to go?
> > > > > 
> > > > > 
> > > > > It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.
> > > > 
> > > > 
> > > > About the MCUmgr client, I don't see it here now: https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html Maybe is missing a `defgroup` or `addtogroup` on docs?
> > > 
> > > 
> > > @jheiskan81 please address this
> > > > In any case, this just shows the hierarchy of groups on doxygen, why would the way the APIs are grouped for documentation not work to show their versions? Moving forward, as more `@version` and `@since` tags are added to doxygen docs, the more complete the table will be, and hopefully, the less "awful" it'll look.
> > > 
> > > 
> > > Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)
> > 
> > 
> > @nordicjm & @edersondisouza #65289 Please check is this fixing a problem?
> 
> Yes, resolves that problem. @edersondisouza now with that, there will be an additional 3 entries, MCUmgr will have 13 entries when it needs 2!

The more the merrier =D

On a serious note, I think this is a valid concern, and there are probably other API/stuff being overexposed now. But we need some convention to be able to identify them (starting with `Z_` or `_`, for instance), and this convention need to be applied. As `Z_` seems too "Zephyr internal-ish", I'd prefer `_`, but would need to fix USB-C as well.

I can update this PR to do just that, and after it's in, people could revisit doxygen groups to fix visibility. Thoughts?

edersondisouza commented:
v11:
  - Ignore groups started with `_`;
  - Remove leading underscore from some USB doxgen groups (they don't seem to be hidden, if that's the idea, this patch can be reverted);
  - Rebased.

edersondisouza commented:
Build failure seems unrelated, fix at https://github.com/zephyrproject-rtos/zephyr/pull/65941

nordicjm commented:
Still way too many MCUmgr groups. I haven't read the thread to see if there is a discussion about the version numbering but why was it picked? How do you update a version? What does the version reference i.e. how to tell if something is experimental, stable, retired, etc.?

edersondisouza commented:
> Still way too many MCUmgr groups. I haven't read the thread to see if there is a discussion about the version numbering but why was it picked? How do you update a version? What does the version reference i.e. how to tell if something is experimental, stable, retired, etc.?

The meaning of version is explained at https://github.com/zephyrproject-rtos/zephyr/blob/aca7fdafb2251b529d1cd2dac5367c4ea3fccc67/doc/develop/api/overview.rst?plain=1. The tl;dr is that for this PR, 0.1 is experimental, 0.8 is unstable and 1.0 is stable, replacing previous wording. To change it, it's just change the `version` tag on the doxygen.

About the MCUmgr groups, there's a way to hide them from this table, which is by having their names starting with `_` (or `z_`). I can patch the MCUmgr in this PR to show the result - then other APIs can be updated later.

nordicjm reviewed (dismissed):
> About the MCUmgr groups, there's a way to hide them from this table, which is by having their names starting with _ (or z_). I can patch the MCUmgr in this PR to show the result - then other APIs can be updated later.

This is absolutely a no.

edersondisouza commented:
> > About the MCUmgr groups, there's a way to hide them from this table, which is by having their names starting with _ (or z_). I can patch the MCUmgr in this PR to show the result - then other APIs can be updated later.
> 
> This is absolutely a no.

Could you please elaborate why? Using the name convention already applies for other doxygen symbols, what's the problem about groups also having them?

nashif commented:
> -- MCUmgr 		
> ---- MCUmgr callback API 		
> ------ MCUmgr fs_mgmt callback API 		
> ------ MCUmgr img_mgmt callback API 		
> ------ MCUmgr os_mgmt callback API 		
> ------ MCUmgr settings_mgmt callback API 		
> ---- MCUmgr handler API 		
> ---- MCUmgr img_mgmt API 		
> ---- MCUmgr mgmt API 	1.0.0 	1.11
> ---- MCUmgr transport SMP API

@nordicjm 

> Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)

that is funny when each group here ends with a big fat **API**

putting all of this aside, this same list is here https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html, so if those are no APIs, they should not be advertised as such or they should be moved to some internal category or hidden using some doxygen magic.

nashif commented:
> About the MCUmgr groups, there's a way to hide them from this table, which is by having their names starting with `_` (or `z_`). I can patch the MCUmgr in this PR to show the result - then other APIs can be updated later.

no, please lets not go there, if something is available via doxygen as an API, it belongs here as well, we should not show it somewhere and hide somewhere else. If some APIs are not to be used by applications etc, they need to be marked as such.

nashif commented:
> It makes sense for how they are laid out in the documentation, it does not make sense at all from an API version perspective.

we can also live with one single version for the top level  group and the version would apply to all sub-groups.

nashif commented:
> * Skip groups starting in `Z_`, as those shall be internal;

We should not be doing something like this here, we need to use the official way of excluding some groups or APIs as internal and AFAIK there are many ways where this can be done. This empty group should be a group at all if it only includes some private functions, I would remove that in some other PR. This PR is definitely not about changing hierachies and defining what internal and what is public.

nordicjm commented:
> > -- MCUmgr
> > ---- MCUmgr callback API
> > ------ MCUmgr fs_mgmt callback API
> > ------ MCUmgr img_mgmt callback API
> > ------ MCUmgr os_mgmt callback API
> > ------ MCUmgr settings_mgmt callback API
> > ---- MCUmgr handler API
> > ---- MCUmgr img_mgmt API
> > ---- MCUmgr mgmt API 	1.0.0 	1.11
> > ---- MCUmgr transport SMP API
> 
> @nordicjm
> 
> > Because they're not APIs, this is not how MCUmgr works, they are services that a remote device e.g. a mobile phone communicates with, they're not something someone calls in an app (i.e. because there are no functions to call)
> 
> that is funny when each group here ends with a big fat **API**
> 
> putting all of this aside, this same list is here https://docs.zephyrproject.org/latest/doxygen/html/group__mcumgr.html, so if those are no APIs, they should not be advertised as such or they should be moved to some internal category or hidden using some doxygen magic.

Which they should be, that has the defines that you can then take and use in your software running on your e.g. pc or phone which has the error code defines and names and command ID defines - both of which were used to populate the values in this external code base here: https://github.com/thedjnK/AuTerm/blob/main/plugins/mcumgr/smp_group_img_mgmt.cpp#L36 and here: https://github.com/thedjnK/AuTerm/blob/main/plugins/mcumgr/smp_group_img_mgmt.cpp#L61

They should not have "API version" entries, just doxygen pages with the defines listed which can be linked to and searched for

nashif commented:
> They should not have "API version" entries, just doxygen pages with the defines listed which can be linked to and searched for

hmm, this is confusing. What is the API status for each of thos APIs (stble, unstable, etc.)? Who defines those API and what type of contract exists for such APIs? Let's say we need to change one of those API calls to fix an issue, do we have to go through the "stable API chanage" process? If there is a change in any of those APIs, breaking or not, will this impact "users" in some way? In short, what promise do we make about those APIs?

nordicjm commented:
> > They should not have "API version" entries, just doxygen pages with the defines listed which can be linked to and searched for
> 
> hmm, this is confusing. What is the API status for each of thos APIs (stble, unstable, etc.)? Who defines those API and what type of contract exists for such APIs? Let's say we need to change one of those API calls to fix an issue, do we have to go through the "stable API chanage" process? If there is a change in any of those APIs, breaking or not, will this impact "users" in some way? In short, what promise do we make about those APIs?

It's not an API as such, it is implementations of the protocol. So looking at the fs mgmt one here: https://docs.zephyrproject.org/latest/services/device_mgmt/smp_groups/smp_group_8.html it implements these, the end user doesn't need to know *how* it does that, just that it does that, because the whole system is transparent to them. What they do need to know is the possible error codes that it can return so they can decode those (on their non-zephyr side, e.g. mobile phone app) and the command IDs so that they can call the right commands. From their zephyr app, they don't need to know anything about the FS mgmt module or call any FS mgmt functions. There is no stability rating for them because it's not something the user does anything with or has any access to from e.g. their code, it's all internal FS mgmt module and MCUmgr protocol code.

edersondisouza commented:
v12:
  - No more selecting of what appears on listing based on naming conventions (or, `_` and `z_` are not ignored anymore). If one wants to hide an API from the listing, one can simply wrap that in the already used `@cond INTERNAL_HIDDEN`.
  - doxmlparser now comes from PyPI instead of TestPyPI, as there was a doxygen release.

thedjnK commented:
> No more selecting of what appears on listing based on naming conventions (or, _ and z_ are not ignored anymore). If one wants to hide an API from the listing, one can simply wrap that in the already used @cond INTERNAL_HIDDEN.

Will this prevent it appearing on doxygen pages or linking to hidden members i.e. enums and structs?

edersondisouza commented:
> > No more selecting of what appears on listing based on naming conventions (or, _ and z_ are not ignored anymore). If one wants to hide an API from the listing, one can simply wrap that in the already used @cond INTERNAL_HIDDEN.
> 
> Will this prevent it appearing on doxygen pages or linking to hidden members i.e. enums and structs?

Nope. If, for instance, one wraps `\defgroup dac_interface` with the conditional, the "dac.h" doxygen file will still be generated and it can be linked from rst files.

edersondisouza commented:
v13:
  - Rebased.

Any comments, @nordicjm and @thedjnK ?

alexanderwachter commented on include/zephyr/drivers/can.h:
```diff
@@ -30,6 +30,8 @@ extern "C" {
 /**
  * @brief CAN Interface
  * @defgroup can_interface CAN Interface
+ * @since 1.14
```
I thought I added CAN in 1.13 🤔

edersondisouza replied:
I just got it from here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/overview.rst?plain=1#L62

:eyes:

henrikbrixandersen replied:
Whoever added that entry had it wrong. The initial CAN driver API was added in 9e12807a6b59816ad0f86e7f4b3b382a02c767f7, which is present even in v1.12.

carlescufi commented:
@gmarull @nordicjm please take another look

@kartben could you please review this one?

gmarull commented:
> @gmarull @nordicjm please take another look

I think this is a flawed idea, not going to approve.

nashif commented:
DNM while people revisit...

nashif commented:
I will spend some time refining the docs for this, stay tuned.

nordicjm commented:
In the current output, the list is just unusable, who's going to look at this https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html through 625 rows of APIs to find an API to use?

henrikbrixandersen commented:
> In the current output, the list is just unusable, who's going to look at this https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html through 625 rows of APIs to find an API to use?

I agree. Having this list in our documentation adds no real, user-oriented value to the project.

gmarull commented:
> In the current output, the list is just unusable, who's going to look at this https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html through 625 rows of APIs to find an API to use? The existing documentation has 94 so the table size has jumped 7 times

Indenting with `-`

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/66de227a-21ad-40e9-8f7e-ada3fd26340d)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/6bbcec0d-608c-463b-8181-1a72ded2ef91)

gmarull commented:
> In the current output, the list is just unusable, who's going to look at this https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html through 625 rows of APIs to find an API to use? The existing documentation has 94 so the table size has jumped 7 times

Versions also seem to be part of the (0.1.0, 0.8.0 and 1.0.0) set only. I am not sure how has this even decided. This doesn't make any sense.

nashif commented:
> Indenting with `-`

yes, not my favorite as well, probably there is a better way to display a tree, suggestions welcome. Change to \_ now, looks a bit better.

nashif commented:
> In the current output, the list is just unusable, who's going to look at this [builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html](https://builds.zephyrproject.io/zephyr/pr/61994/docs/develop/api/overview.html) through 625 rows of APIs to find an API to use? The existing documentation has 94 so the table size has jumped 7 times

"It is what it is" This is how many APIs we have. If you are interested in 1 API, you can search for it in the table. The table loads juist fine for me.
The fact that the existing list was short is because it was unmaintained and incomplete. This is what we are trying to sovle with this PR among other things.

jukkar commented on include/zephyr/net/net_core.h:
```diff
@@ -29,6 +29,8 @@ extern "C" {
 /**
  * @brief Networking
  * @defgroup networking Networking
+ * @since 1.0
+ * @version 1.0.0
```
I don't think this should have any version information, as it is an umbrella entry without anything in the group. All the other network related groups are children of this "networking" group but it does not make them stable. If one looks at the generated documentation list, it gives an impression that all the network APIs are v1.0.0 which is not the case.

kartben commented:
> > Indenting with `-`
> 
> yes, not my favorite as well, probably there is a better way to display a tree, suggestions welcome. Change to _ now, looks a bit better.

as a quick hack, I am guessing `&nbsp` would achieve the same goal without being visible

nashif commented:
> as a quick hack, I am guessing `&nbsp` would achieve the same goal without being visible

I tried this, did not work.

carlescufi reviewed (changes requested):
The overall idea and implementation is fine. I am not particularly a fan of versioning individual APIs but I'll take it as a means to help our users.

But I completely disagree that we should sacrifice the cleanliness of the HTML build to improve the .pdf one. I think the HTML one should come first, and those `_` are just very distracting and reflect poorly on the quality of our doc. I suggest @edersondisouza  reverts back to spaces.

nashif commented:
> But I completely disagree that we should sacrifice the cleanliness of the HTML build to improve the .pdf one. I think the HTML one should come first, and those `_` are just very distracting and reflect poorly on the quality of our doc. I suggest @edersondisouza reverts back to spaces.

yes, I am also fine with that, the indentation and table looks bad with the '_'

edersondisouza commented:
Ok, found a way to indent with spaces that seems to work both with HTML and PDF.

carlescufi reviewed (approved):
LGTM now, thank you!
